woocommerce/google-listings-and-ads
# Import from an externalized package if explicitly asked for

This is an alternative/complimentary PR to https://github.com/woocommerce/google-listings-and-ads/pull/1739
### Changes proposed in this Pull Request:

- [Import from an externalized package if explicitly asked for,](https://github.com/woocommerce/google-listings-and-ads/commit/a57b0be588dd5c0c4dc4ca33b7b1bd10594401e1) 
   even if the package is bundled as well.
   So we could selectively externalize specific imports and gradually externalize all cases.

- [Import ExternalLink from DEWPed @wordpress/components,](https://github.com/woocommerce/google-listings-and-ads/commit/ad10229ada8c08d6542ac66260bee5b24a26242e) 
   even though we bundle that package.
   Fixes non-working `VisuallyHidden` styles for Google Ads settings https://github.com/woocommerce/google-listings-and-ads/issues/1736,
  and billing setup.


### Screenshots:

![2022-10-20 18 21 57](https://user-images.githubusercontent.com/17420811/196924232-d11c3562-e9ef-4ba5-a667-2efe04fb8c06.png)
![image](https://user-images.githubusercontent.com/17435/201194740-f4e23536-e86a-4fac-a927-b7b5e57764f0.png)


### Detailed test instructions:


1. Checkout
2. `npm i && npm run build`
3. Go to paid ads setup `/wp-admin/admin.php?page=wc-admin&path=%2Fgoogle%2Fsetup-ads` with non-approved billing
4. Check The external links  on step "1 Set up your accounts" & "3 Set up billing"


### Additional details:
1. I'm using `@wordpress/dewp` implementation details here. This could be unstable. If we'd like to move on with this change, I'd propose a change upstream to improve DEWP to be more extendable. as IMHO, currently, it's not, that's why `@woocommerce/DEWP` needs to extend the class, and more-or-less duplicate `@wordpress`es code. 

   I believe `@woocommerce/DEWP` and our case here could be simplified if the exposed API, would allow doing something like:
```js
requestToExternal = ( request, defaultAction ) => {
	if ( request.startsWith( explicitlyExtractPrefix ) ) {
		 defaultAction( request.substr( explicitlyExtractPrefix.length ) )
   }
   // ...
}
```
or
```js
requestToExternal = ( request ) => {
	if ( request.startsWith( explicitlyExtractPrefix ) ) {
		 DependencyExtractionWebpackPlugin.requestToExternal( request.substr( explicitlyExtractPrefix.length ) )
   }
   // ...
}
```

2. Currently, this setup supports only packages extracted by `@wordpress/DEWP`, as `@woo…`'s one does not export its `requestToExternal`. So we would have to duplicate its code, which IMHO is even worse than the plugin into implementation details like `@…/dependency-extraction-webpack-plugin/lib/util`. I believe that could be solved in a cleaner fashion if the proposal from the point above gets accepted. Plus, being able to do so just for `@wordpress/` package is already a value.
3. Alternatively, we could explicitly bundle, instead of extract. Then the implementation would not face the problems of the points above, as we'd just do the current opt-out selectively.
   However, that would require us to prefix every import we have with `bundled/` which may bring lots of noise.
4. ~I added `eslint-import-resolver-alias` as I didn't manage to configure an alias for `node_modules` with `eslint-import-resolver-webpack`. I think ideally, we should stick to one. As we already have quite many places responsible for setting aliases: `webpack.config.js`, `js/src/jsconfig.json`, `jest.config.js`, now `.eslintrc.js`. I'd appreciate if somebody could clean this up.~ solved in https://github.com/woocommerce/google-listings-and-ads/pull/1762#issuecomment-1311438333
5. [Externalized report](https://github.com/woocommerce/google-listings-and-ads/pull/1762/files#diff-2ccc451e09095e78419f3f696655eda1da857eb2213c97996783d2accc6f042c) shows `extracted/@wordpress/components` as and externalized package. With more extendable DEWP, that exposes its primitives instead of simplifying everything to returning `undefined` on path, we could avoid that by doing something like:
```js
requestToExternal = ( request ) => {
	if ( request.startsWith( explicitlyExtractPrefix ) ) {
		 return externalizeWpDeps( request.substr( explicitlyExtractPrefix.length ) )
    }
   // filter bundled opt-outs.
```


### Changelog entry

> Fix - WordPress 5.9 Compatibility: Visually hide descriptions of external link icons.
> Fix - WordPress 6.1 Compatibility: Popover and Tooltip components should be displayed as floating.
> Dev - Selectively externalize bundled packages.
